### PR TITLE
Generate and parse links with the session and user keys in the url

### PR DESCRIPTION
**Pull request by Ra1issa**
1. Analyst generates links from /trusted/session_data.html using a server API.
2. Clients open the links and use them to submit data to the server.
3. Server checks whether submitted user key and session key match.

### CODE REVIEW

**Comment by KinanBab** [comment]
Added 2 extra commits:
1. add password to a session, and require analyst to type in password before being able to unmask/view/generate links.
2. small fix in server code (a new variable had the same name as an old one).

Maybe hash the password before sending it to server?

File: client/index.html [hunk 1]
```diff
@@ -197,8 +195,20 @@ <h3 id="totals-hot-name"></h3>
 
     </div>
   <br/>
-  <script type="text/javascript"> 
+  <script type="text/javascript">
     $(document).ready(function() {
+
+      //Copied from trusted/session_data
+      window.getParameterByName = function(name) {
```
Comment (frederickjansen) [inline client/index.html:202]: Let's avoid creating global functions when possible. I don't think it's necessary here.

File: client/index.html [hunk 2]
```diff
@@ -197,8 +195,20 @@ <h3 id="totals-hot-name"></h3>
...
+        name = name.replace(/[\[]/, "\\[").replace(/[\]]/, "\\]");
+        var regex = new RegExp("[\\?&]" + name + "=([^&#]*)"),
```
Comment (frederickjansen) [inline client/index.html:204]: Do we really need regex for this? Parsing the query string is probably more readable with a simple split.

**Comment by frederickjansen** [comment]
Some general comments while I'm reviewing the code:
- I'm going to introduce eslint (hopefully soon). It's clear that not everyone is using the same IDE, and the amount of changes due to whitespace is way too high.
- Can we check the user key on page load? If a user gets it wrong for whatever reason, there's no way to tell until data submission. At that point there's no way to fix it without reloading the page and starting over again. A separate input box would also be possible.
- General UI stuff, we should polish the interface.

File: trusted/index.html [hunk 3]
```diff
@@ -31,21 +31,23 @@
...
+        All secured data will be lost if the private key is lost. Also, do <b>not</b> share your private key. After clicking the
+        "Generate Session" button, email the <b>Session Key</b> to all participants. Once the data is collected from the
         participants, with your private key, continue to the next step.
       </p>
     </div>
     <br/>
-    <button class="btn btn-primary btn-lg" id="Generate" 
-            onclick="generateSession('infoDiv','sessionID','pubkeyID','privkeyID','linkID')">
+    <button class="btn btn-primary btn-lg" id="Generate"
+            onclick="generateSession('infoDiv','sessionID', 'passwordID', 'pubkeyID','privkeyID','linkID')">
```
Comment (frederickjansen) [inline trusted/index.html:41]: Save the password to a file as well as the private key.

File: trusted/session_data.html [hunk 4]
```diff
@@ -31,9 +41,12 @@
     </div>
     <form class="form-inline" role="form" id="inputform">
       <input type="text" id="sess" placeholder="Session Number"/>
-      <button id="submit" type="button" class="btn btn-primary" onclick="genT()" >SUBMIT</button>
+      <input type="password" id="pass" placeholder="Session Password"/>
+      <button id="submit" type="button" class="btn btn-primary" onclick="genT()">SUBMIT</button>
```
Comment (frederickjansen) [inline trusted/session_data.html:45]: I know this code was there before, but we should be more consistent about our coding style. Onclick in html, document.getElementById jQuery, ...

**Pull request merged**